### PR TITLE
Weight bit-accuracy of Dense (Python vs. Loihi 2)

### PR DESCRIPTION
**Pull request by mathisrichter**
Issue Number:

Objective of pull request: Public PR that goes hand-in-hand with the PR in the Lava-on-Loihi extension.

## Pull request checklist

Your PR fulfills the following requirements:
- [ ] [Issue](https://github.com/lava-nc/lava/issues) created that explains the change and why it's needed
- [ ] Tests are part of the PR (for bug fixes / features)
- [ ] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [ ] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [ ] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [ ] Lint (`flakeheaven lint src/lava tests/`) and (`bandit -r src/lava/.`) pass locally
- [ ] Build tests (`pytest`) passes locally

## Pull request type

 

 

Please check your PR type:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Review by bamsumit — APPROVED** [review]

**Comment by bamsumit** [comment]
Looks good overall Mathis. Haven't looked at it in detail.

**Review by mgkwill — APPROVED** [review]

File: src/lava/utils/weightutils.py [hunk 1]
```diff
@@ -17,81 +18,154 @@ class SignMode(Enum):
     INHIBITORY = 3
 
 
-def optimize_weight_bits(weight: np.ndarray,
-                         loihi2: bool = False) -> ty.Tuple[np.ndarray,
-                                                           int,
-                                                           int,
-                                                           SignMode]:
+def determine_sign_mode(weights: np.ndarray) -> SignMode:
```
Comment (weidel-p) [inline src/lava/utils/weightutils.py:21]: Please add a docstring

File: src/lava/utils/weightutils.py [hunk 2]
```diff
@@ -17,81 +18,154 @@ class SignMode(Enum):
...
+        if sign_mode == SignMode.MIXED:
+            weights = weights // 2
+
+    optimized_weights = OptimizedWeights(weights=weights.astype(int),
+                                         num_weight_bits=num_weight_bits,
+                                         weight_exp=weight_exp)
+    return optimized_weights
+
+
+def _determine_weight_exp(weights: np.ndarray,
```
Comment (weidel-p) [inline src/lava/utils/weightutils.py:83]: Please add a docstring

File: src/lava/utils/weightutils.py [hunk 3]
```diff
@@ -17,81 +18,154 @@ class SignMode(Enum):
...
+    scale_bits = int(np.floor(np.log2(scale)))
+    if sign_mode == SignMode.MIXED:
+        scale_bits += 1
 
+    weight_exp = -scale_bits
+
+    return weight_exp
+
+
+def _determine_num_weight_bits(weights: np.ndarray,
```
Comment (weidel-p) [inline src/lava/utils/weightutils.py:108]: Please add a docstring

File: src/lava/utils/weightutils.py [hunk 4]
```diff
@@ -17,81 +18,154 @@ class SignMode(Enum):
...
+
+    # Saturate the weights according to the sign_mode.
+    mixed_flag = int(sign_mode == SignMode.MIXED)
+    excitatory_flag = int(sign_mode == SignMode.EXCITATORY)
+    inhibitory_flag = int(sign_mode == SignMode.INHIBITORY)
+
+    min_wgt = (-2 ** 8) * (mixed_flag + inhibitory_flag)
+    max_wgt = (2 ** 8 - 1) * (mixed_flag + excitatory_flag)
+
+    clipped_weights = np.clip(weights, min_wgt, max_wgt)
```
Comment (weidel-p) [inline src/lava/utils/weightutils.py:165]: It seems this function not only truncates the least significant bits but also clips the weights. I would prefer to have a separate function for that or at least mention that in the docstring

**Review by weidel-p — APPROVED** [review]
I think this PR fixes not only a bug, but also makes the code look cleaner. The only request I have is to add docstrings to all functions. Nice amount of tests.

File: src/lava/utils/weightutils.py (hunk 1 above)
Comment (mathisrichter) [reply src/lava/utils/weightutils.py:20]: Addressed in https://github.com/lava-nc/lava/pull/328/commits/7476abe0ebe64dd4fb79c079d553955a2f4f94fd

File: src/lava/utils/weightutils.py (hunk 2 above)
Comment (mathisrichter) [reply src/lava/utils/weightutils.py:94]: Addressed in https://github.com/lava-nc/lava/pull/328/commits/7476abe0ebe64dd4fb79c079d553955a2f4f94fd

File: src/lava/utils/weightutils.py (hunk 3 above)
Comment (mathisrichter) [reply src/lava/utils/weightutils.py:134]: Addressed in https://github.com/lava-nc/lava/pull/328/commits/7476abe0ebe64dd4fb79c079d553955a2f4f94fd

File: src/lava/utils/weightutils.py (hunk 4 above)
Comment (mathisrichter) [reply src/lava/utils/weightutils.py:240]: I separated the clipping into its own function and wrote unit tests. See https://github.com/lava-nc/lava/pull/328/commits/7476abe0ebe64dd4fb79c079d553955a2f4f94fd

**Review by gkarray — APPROVED** [review]
Everything looks good, nice work Mathis!
I especially like the fact of removing Vars for `weight_exp`, `num_weight_bits` and `sign_mode` and passing them through the `proc_params` argument instead.

**Pull request merged**